zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/usermanagement/UserController.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.gateway.rest.controller.usermanagement;
+
+import io.camunda.identity.usermanagement.CamundaUser;
+import io.camunda.identity.usermanagement.CamundaUserWithPassword;
+import io.camunda.identity.usermanagement.service.UserService;
+import io.camunda.zeebe.gateway.protocol.rest.CamundaUserDto;
+import io.camunda.zeebe.gateway.protocol.rest.CamundaUserWithPasswordDto;
+import io.camunda.zeebe.gateway.protocol.rest.SearchRequestDto;
+import io.camunda.zeebe.gateway.protocol.rest.UserSearchResponseDto;
+import io.camunda.zeebe.gateway.rest.controller.ZeebeRestController;
+import java.util.List;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ZeebeRestController
+@RequestMapping("/v2/users")
+public class UserController {
+  private final UserService userService;
+
+  public UserController(final UserService userService) {
+    this.userService = userService;
+  }
+
+  @PostMapping(
+      produces = {MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_PROBLEM_JSON_VALUE},
+      consumes = MediaType.APPLICATION_JSON_VALUE)
+  @ResponseStatus(HttpStatus.CREATED)
+  public CamundaUserDto createUser(
+      @RequestBody final CamundaUserWithPasswordDto userWithPasswordDto) {
+    return mapToCamundaUserDto(userService.createUser(mapToUserWithPassword(userWithPasswordDto)));
+  }
+
+  @DeleteMapping(path = "/{id}")
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  public void deleteUser(@PathVariable final Long id) {
+    userService.deleteUser(id);
+  }
+
+  @GetMapping(
+      path = "/{id}",
+      produces = {MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_PROBLEM_JSON_VALUE})
+  public CamundaUserDto findUserByUsername(@PathVariable final Long id) {
+    return mapToCamundaUserDto(userService.findUserById(id));
```

Comment:
💭 Without explicit exception handling, you'll get the default problem detail creation provided by the global advice. If you want to steer the different HTTP status codes for different exceptions, you'll have to handle exceptions yourself and rather use a generic `ResponseEntity` return type. Feel free to have a look at the other controllers for examples.

Reply:
The things we're validating at the moment all refer to invalid input, thus I moved them to be IllegalArgumentExceptions which will be picked up by the global handler and return the appropriate BAD_REQUEST for now.

I've created an issue for us to revisit the exception handling once we've got all new user management endpoints in place / with the last bunch. This way we can make it consistent as - supposedly - there will be more corner cases to handle.

Reply:
We do input validation in other controllers as well, e.g. in the `UserTaskController`.

It validates the input using the `RequestMapper`, e.g. [for user task update requests](https://github.com/camunda/camunda/blob/main/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RequestMapper.java#L107-L124). Those mapping methods return an `Either` that represents either a failure (`Either.Left`) or success (`Either.Right`). If there is a validation error, we return a `ProblemDetail` (`Either.Left`). If it's fine, we return a broker request object (`Either.Right`).

In the controller, we either send the broker request to the broker (that is later mapped to a `ResponseEntity` once we get the result) or map the problem detail directly to a respective `ResponseEntity` with a 4xx HTTP code. That all happens using the `fold` method of the `Either` which simply says "if left exists, transform and return that, otherwise, transform and return right".

Happy to walk you through it synchronously, if you like.

Reply:
If you ask me, it's simpler to grasp the controllers if it's obvious how they return responses. If this all happens transparently at some point because the global exception handlers kick in, the controllers are harder to understand. And harder to test, actually. Because you always need the full Spring application to instantiate the global handlers to test exceptional cases. This is not necessary if the controllers return responses directly themselves. This makes the test classes less prone to adjustments because they are more isolated.

Reply:
I would also argue to not postpone this to a later point when there are even more endpoints to adjust then. The required changes are not big right now I think, happy to assist.

Reply:
I would also argue to not postpone this to a later point when there are even more endpoints to adjust then. The required changes are not big right now I think, happy to assist.

Reply:
I've added explicit exception handling to the controller methods as discussed yesterday.

Reply:
🚀